### PR TITLE
[FLINK-30620] Table Store Hive catalog supports specifying custom Hive metastore client

### PR DESCRIPTION
**Pull request by tsreaper**
Currently Hive metastore client class is hard coded in HiveCatalog, however users may want to specify custom Hive metastore client to read and write Hive compliant storage.

This PR is tested by hand.

### CODE REVIEW

File: docs/layouts/shortcodes/generated/hive_catalog_configuration.html [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+            <th class="text-left" style="width: 20%">Key</th>
+            <th class="text-left" style="width: 15%">Default</th>
+            <th class="text-left" style="width: 10%">Type</th>
+            <th class="text-left" style="width: 55%">Description</th>
+        </tr>
+    </thead>
+    <tbody>
+        <tr>
+            <td><h5>metastore.client.class</h5></td>
+            <td style="word-wrap: break-word;">"org.apache.flink.table.store.shaded.org.apache.hadoop.hive.metastore.HiveMetaStoreClient"</td>
```
Comment (JingsongLi) [inline docs/layouts/shortcodes/generated/hive_catalog_configuration.html:13]: Maybe we don't need this documentation. I think this option should be inner.
It is not so good to expose a shaded interface. (`IMetaStoreClient` is shaded)

File: flink-table-store-hive/flink-table-store-hive-catalog/src/main/java/org/apache/flink/table/store/hive/HiveCatalog.java [hunk 2]
```diff
@@ -63,6 +64,9 @@
 /** A catalog implementation for Hive. */
 public class HiveCatalog extends AbstractCatalog {
 
+    public static final String INTERNAL_METASTORE_CLIENT_CLASS =
```
Comment (JingsongLi) [inline flink-table-store-hive/flink-table-store-hive-catalog/src/main/java/org/apache/flink/table/store/hive/HiveCatalog.java:67]: Maybe we don't need this key.
We can just pass class by the constructor of `HiveCatalog`.

File: flink-table-store-hive/flink-table-store-hive-catalog/src/main/java/org/apache/flink/table/store/hive/HiveCatalogOptions.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import org.apache.hadoop.hive.metastore.HiveMetaStoreClient;
+
+/** Specific config options for {@link HiveCatalogFactory}. */
+public class HiveCatalogOptions {
+
+    public static final ConfigOption<String> METASTORE_CLIENT_CLASS =
+            ConfigOptions.key("metastore.client.class")
+                    .stringType()
+                    .defaultValue(HiveMetaStoreClient.class.getName())
+                    .withDescription("Class name of Hive metastore client");
```
Comment (JingsongLi) [inline flink-table-store-hive/flink-table-store-hive-catalog/src/main/java/org/apache/flink/table/store/hive/HiveCatalogOptions.java:33]: NOTE it should be shaded?

File: flink-table-store-hive/flink-table-store-hive-catalog/src/main/java/org/apache/flink/table/store/hive/HiveCatalogFactory.java [hunk 4]
```diff
@@ -41,13 +40,21 @@ public String identifier() {
...
+                        options.get(CatalogOptions.URI),
+                        CatalogOptions.URI.key()
+                                + " must be set for table store "
+                                + IDENTIFIER
+                                + " catalog");
+
         org.apache.hadoop.conf.Configuration hadoopConfig =
                 new org.apache.hadoop.conf.Configuration();
         options.toMap().forEach(hadoopConfig::set);
+        hadoopConfig.set(
```
Comment (JingsongLi) [inline flink-table-store-hive/flink-table-store-hive-catalog/src/main/java/org/apache/flink/table/store/hive/HiveCatalogFactory.java:52]: We can add test for here. May a mock `IMetaStoreClient` just implement `getAllDatabases` to verify.

File: flink-table-store-hive/flink-table-store-hive-catalog/src/main/java/org/apache/flink/table/store/hive/HiveCatalogOptions.java (hunk 3 above)
Comment (tsreaper) [reply flink-table-store-hive/flink-table-store-hive-catalog/src/main/java/org/apache/flink/table/store/hive/HiveCatalogOptions.java:33]: Not necessary. If a user has his own client, it is very likely not shaded. Only our default client is shaded.

Comment (JingsongLi) [reply flink-table-store-hive/flink-table-store-hive-catalog/src/main/java/org/apache/flink/table/store/hive/HiveCatalogOptions.java]: `HiveMetaStoreClient` have been shaded too.

File: docs/content/docs/how-to/creating-catalogs.md [hunk 5]
```diff
@@ -31,6 +31,8 @@ Table Store catalogs currently support two types of metastores:
 * `filesystem` metastore (default), which stores both metadata and table files in filesystems.
 * `hive` metastore, which additionally stores metadata in Hive metastore. Users can directly access the tables from Hive.
 
+See [CatalogOptions]({{< ref "docs/maintenance/configurations#catalogoptions" >}}) for detailed options when creating a catalog.
```
Comment (JingsongLi) [inline docs/content/docs/how-to/creating-catalogs.md:34]: remove this?

File: flink-table-store-hive/flink-table-store-hive-catalog/src/main/java/org/apache/flink/table/store/hive/HiveCatalog.java [hunk 6]
```diff
@@ -408,12 +411,16 @@ private Lock lock(ObjectPath tablePath) {
...
     }
 
-    static IMetaStoreClient createClient(HiveConf hiveConf) {
+    static IMetaStoreClient createClient(HiveConf hiveConf, String clientClassName) {
         IMetaStoreClient client;
         try {
             client =
                     RetryingMetaStoreClient.getProxy(
-                            hiveConf, tbl -> null, HiveMetaStoreClient.class.getName());
+                            hiveConf,
```
Comment (JingsongLi) [inline flink-table-store-hive/flink-table-store-hive-catalog/src/main/java/org/apache/flink/table/store/hive/HiveCatalog.java:419]: We can verify multiple hive versions. I don't know if it is good to all supported versions.

File: docs/content/docs/how-to/creating-catalogs.md (hunk 5 above)
Comment (tsreaper) [reply docs/content/docs/how-to/creating-catalogs.md:34]: I think this is a necessary information. This change is not related to Hive.

File: flink-table-store-hive/flink-table-store-hive-catalog/src/main/java/org/apache/flink/table/store/hive/HiveCatalog.java (hunk 6 above)
Comment (tsreaper) [reply flink-table-store-hive/flink-table-store-hive-catalog/src/main/java/org/apache/flink/table/store/hive/HiveCatalog.java:422]: I've tested Hive 2.1, 2.2, 2.3 and 3.1. They are all supported.

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**